### PR TITLE
Improve cobertura absolute/relative path report generation

### LINKED ISSUE #413: Unable to change paths in cobertura report
Getting an issue where our application is compiled within docker and we're trying to export the cobertura xml file from it. 

Since the `filename` attribute is the full path to the file, we need to do a search a replace. 

Adding the root of the project to `sources/source` and using relative paths for `filename`, it would be a lot easier to manager.

### PR DESCRIPTION
**Pull request by daveMueller**
closes #413

### CODE REVIEW

**Comment by daveMueller** [comment]
I've tested it on Windows and Linux both with Jenkins build server and Cobertura Plugin. ReportGenerator is also still working.

File: src/coverlet.core/Reporters/CoberturaReporter.cs [hunk 1]
```diff
@@ -32,8 +32,8 @@ public string Report(CoverageResult result)
             coverage.Add(new XAttribute("timestamp", (int)(DateTime.UtcNow - new DateTime(1970, 1, 1)).TotalSeconds));
 
             XElement sources = new XElement("sources");
-            var rootDirs = GetRootDirs(result.Modules, result.UseSourceLink).ToList();
-            rootDirs.ForEach(x => sources.Add(new XElement("source", x)));
+            var basePaths = GetBasePaths(result.Modules, result.UseSourceLink).ToList();
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Reporters/CoberturaReporter.cs]: maybe better change the name from `basePaths` to `absolutePaths`  it's more clear.

File: test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs [hunk 2]
```diff
@@ -146,43 +146,65 @@ public void TestReportWithTwoDifferentDirectories()
 
             string absolutePath1;
             string absolutePath2;
+            string absolutePath3;
+            string absolutePath4;
+            string absolutePath5;
+            string absolutePath6;
 
             if (isWindows)
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs]: `isWindows` is used only in one place use directly `RuntimeInformation.IsOSPlatform(OSPlatform.Windows)`

File: test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs [hunk 3]
```diff
@@ -146,43 +146,65 @@ public void TestReportWithTwoDifferentDirectories()
...
             {
-                absolutePath1 = @"C:\projA\file.cs";
-                absolutePath2 = @"E:\projB\file.cs";
+                absolutePath1 = @"C:\projA\dir1\dir10\file1.cs";
```
Comment (MarcoRossignoli) [inline test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs]: add test for `F:\file1.cs`

Comment (MarcoRossignoli) [reply test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs]: and maybe also empty roots `F:\` `/` if there is an error on modules data...error should be clear no out of range or cryptic one.

File: src/coverlet.core/Reporters/CoberturaReporter.cs [hunk 4]
```diff
@@ -133,28 +133,48 @@ public string Report(CoverageResult result)
...
             if (useSourceLink)
             {
                 return new[] { string.Empty };
             }
 
-            return modules.Values.SelectMany(k => k.Keys).Select(Directory.GetDirectoryRoot).Distinct();
+            var pathsGroupedByRootDir = modules.Values.SelectMany(k => k.Keys).GroupBy(Directory.GetDirectoryRoot);
+            var basePaths = pathsGroupedByRootDir.Select(group =>
+            {
+                var splittedPaths = group.Select(x => x.Split(Path.DirectorySeparatorChar)).OrderBy(x => x.Length).ToList();
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Reporters/CoberturaReporter.cs]: rename `x` in `absolutePath` I don't like single letter var name, code is hard to read.

File: src/coverlet.core/Reporters/CoberturaReporter.cs [hunk 5]
```diff
@@ -133,28 +133,48 @@ public string Report(CoverageResult result)
...
+                if (splittedPaths.Count == 1)
+                {
+                    return group.Key;
+                }
+
+                var basePathSegments = new List<string>();
+
+                splittedPaths[0].Select((value, index) => (value, index)).ToList().ForEach(x =>
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Reporters/CoberturaReporter.cs]: ditto, rename `valueIndexPair`

File: src/coverlet.core/Reporters/CoberturaReporter.cs [hunk 6]
```diff
@@ -133,28 +133,48 @@ public string Report(CoverageResult result)
             return Encoding.UTF8.GetString(stream.ToArray());
         }
 
-        private static IEnumerable<string> GetRootDirs(Modules modules, bool useSourceLink)
+        private static IEnumerable<string> GetBasePaths(Modules modules, bool useSourceLink)
         {
             if (useSourceLink)
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Reporters/CoberturaReporter.cs]: Please add some comment on the idea behind alg for instance 

```
/*
     Workflow

     Path1 c:\dir\file1.cs
	 Path2 c:\dir\file2.cs
	 Path3 e:\dir\file2.cs
     

     1) Search for root dir 
		c:\ ->	c:\dir\file1.cs
				c:\dir\file2.cs
		e:\ ->	e:\dir\file2.cs
		
	 2) Split path on directory separator i.e. for record c:\ ordered ascending by fragment elements
		 Path1 = [c:|dir|file1.cs]
		 Path2 = [c:|dir|file2.cs]
		 
	 3)  Find longest shared path comparing indexes, list is ordered so the longest absolute path will be max Path[0] length
		 
		 Path1[0] = Path2[0] -> add to final fragment list
		 Path1[n] = PathY[n] -> add to final fragment list
		 ...
		 
     4) Concat created fragment list
*/
```

File: src/coverlet.core/Reporters/CoberturaReporter.cs [hunk 7]
```diff
@@ -133,28 +133,48 @@ public string Report(CoverageResult result)
...
             {
                 return new[] { string.Empty };
             }
 
-            return modules.Values.SelectMany(k => k.Keys).Select(Directory.GetDirectoryRoot).Distinct();
+            var pathsGroupedByRootDir = modules.Values.SelectMany(k => k.Keys).GroupBy(Directory.GetDirectoryRoot);
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Reporters/CoberturaReporter.cs]: `pathsGroupedByRootDir` is used only in one place use it directly
```cs
           return modules.Values.SelectMany(k => k.Keys).GroupBy(Directory.GetDirectoryRoot).Select(group =>
            {
               ...
```

**Comment by MarcoRossignoli** [comment]
Some comments but idea is ok to me, thank's!

**Comment by MarcoRossignoli** [comment]
FYI after merge of this https://github.com/dotnet/runtime/issues/934#issuecomment-523152939 we'll do splits with zero allocation.

File: src/coverlet.core/Reporters/CoberturaReporter.cs (hunk 1 above)
Comment (daveMueller) [reply src/coverlet.core/Reporters/CoberturaReporter.cs]: Done

File: test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs (hunk 2 above)
Comment (daveMueller) [reply test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs]: Done

File: test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs (hunk 3 above)
Comment (daveMueller) [reply test/coverlet.core.tests/Reporters/CoberturaReporterTests.cs]: Done. Implementation can handle empty roots without throwing.

File: src/coverlet.core/Reporters/CoberturaReporter.cs (hunk 4 above)
Comment (daveMueller) [reply src/coverlet.core/Reporters/CoberturaReporter.cs]: Done

File: src/coverlet.core/Reporters/CoberturaReporter.cs (hunk 5 above)
Comment (daveMueller) [reply src/coverlet.core/Reporters/CoberturaReporter.cs]: Done

File: src/coverlet.core/Reporters/CoberturaReporter.cs (hunk 6 above)
Comment (daveMueller) [reply src/coverlet.core/Reporters/CoberturaReporter.cs]: Done

File: src/coverlet.core/Reporters/CoberturaReporter.cs (hunk 7 above)
Comment (daveMueller) [reply src/coverlet.core/Reporters/CoberturaReporter.cs]: Done

**Comment by abbotware** [comment]
running the nightly build - I noticed the slashes are not normalized 

on linux the paths are all:
folder/folder/file

and windows:
folder\folder\file

Might not be an issue.. but perhaps worth noting

**Comment by MarcoRossignoli** [comment]
> running the nightly build - I noticed the slashes are not normalized

I don't think that's a problem, reporters should understand all OS path format and @daveMueller can confirm that it works on some reporters like jenkins/report generator...btw before this update were they normalized in some way?

**Comment by daveMueller** [comment]
Yes I can confirm that it works on the reporters Jenkins Cobertura Plugin and Report Generator. I checked the old code and there it also wasn't normalized (https://github.com/tonerdo/coverlet/blob/b3c99fffe757431403a0f4f3c4cc1e798ff97863/src/coverlet.core/Reporters/CoberturaReporter.cs#L155).
Also ```Path.DirectorySeparatorChar``` is used to concatenate the strings and thus the format is bound to the OS specific path.

**Pull request merged**